### PR TITLE
Expose ResultCode and DiagnosticMessage on SearchResult

### PR DESCRIPTION
**Pull request by Pujathacker2210**
## Problem

`GetLDAPError` returns `nil` when `resultCode` is 0 (success), discarding
the `diagnosticMessage` from `SearchResultDone`. Per RFC 4511 §4.1.9, servers
may set `diagnosticMessage` even on a successful operation to communicate
additional information such as degraded state.

For example, Red Hat IPA sets `diagnosticMessage` when the directory is
reinitializing but still returns `resultCode` 0 with an empty entry list.
Callers of `Search()` currently have no way to see that message — they only
see zero entries and no error, which is indistinguishable from a genuine
empty result.

## Solution

- Add `ResultCode` and `DiagnosticMessage` fields to `SearchResult`.
- In `Search()`, populate them from the `SearchResultDone` packet (case 5) **before** calling `GetLDAPError`, so they are preserved even when`GetLDAPError` returns `nil` on success.
- Add `ParseLDAPResult` helper that extracts `resultCode`, `matchedDN`, and `diagnosticMessage` from an `LDAPResult` BER packet without the early return on success that `GetLDAPError` has.
- Copy the new fields in `appendTo` so paged searches preserve them.

## Backward compatibility

- `GetLDAPError` is unchanged.
- `SearchResult` is a struct; adding fields is backward compatible in Go.
- Existing callers are unaffected — they never read the new fields.
- Callers who need the diagnostic on success can now read `result.DiagnosticMessage` after `Search()` returns.

## References

- [RFC 4511 §4.1.9](https://datatracker.ietf.org/doc/html/rfc4511#section-4.1.9) — LDAPResult definition (resultCode + diagnosticMessage)
- [RFC 4511 §4.5.2](https://datatracker.ietf.org/doc/html/rfc4511#section-4.5.2) — SearchResultDone is an LDAPResult

### Prior art

Python's `python-ldap` library already exposes the diagnostic message after successful operations by calling `get_option(OPT_DIAGNOSTIC_MESSAGE)` on the LDAP handle ([source](https://github.com/python-ldap/python-ldap/blob/main/Lib/ldap/ldapobject.py#L152)). This PR brings the same capability to `go-ldap` by surfacing `diagnosticMessage` on the `SearchResult` struct regardless of `resultCode`.

### CODE REVIEW

**Comment by Pujathacker2210** [comment]
@t2y  @cpuschma @johnweldon - Request you to review the PR

**Comment by t2y** [comment]
This is my personal opinion based on the changes. Before we proceed with this PR, I'd like others' opinions on two API changes.

1. Add `DiagnosticMessage() string` to the `Response` interface

`SearchAsync` drops the diagnosticMessage on success just like `Search`. To support both, I'd add this method and have it work like `Err()`, i.e. valid after `Next` returns false. This breaks external implementations of `Response` (e.g. test fakes), but fixing them only takes one extra method.

2. Add a `DiagnosticMessage` field to `ldap.Error`

The goal is to parse LDAPResult in one place. `GetLDAPError` and the success path in `Search`/`SearchAsync` would share one internal parser, and `GetLDAPError` would copy the diagnosticMessage into the new field. Callers could then read it as a plain field on both success (`SearchResult.DiagnosticMessage`) and failure (`Error.DiagnosticMessage`), instead of pulling it out of `Err`, which also holds client-side errors.

File: v3/error.go [hunk 1]
```diff
@@ -195,6 +195,46 @@ func (e *Error) Error() string {
...
+// state (for example, a directory in read-only / reinitializing mode).
+func ParseLDAPResult(packet *ber.Packet) (resultCode uint16, matchedDN, diagnosticMessage string) {
+	if packet == nil || len(packet.Children) < 2 || packet.Children[1] == nil {
+		return 0, "", ""
+	}
+	response := packet.Children[1]
+	if len(response.Children) < 1 || response.Children[0] == nil || response.Children[0].Value == nil {
+		return 0, "", ""
+	}
+	switch v := response.Children[0].Value.(type) {
```
Comment (cpuschma) [inline v3/error.go]: TagInteger can only of type int64, no need to check for the other types.

File: v3/error.go [hunk 2]
```diff
@@ -195,6 +195,46 @@ func (e *Error) Error() string {
...
+// ParseLDAPResult extracts resultCode, matchedDN, and diagnosticMessage from a
+// BER packet representing an LDAPResult (e.g. SearchResultDone).
+//
+// Unlike GetLDAPError, ParseLDAPResult does not return early when resultCode is
+// success (0). This allows callers to read the diagnosticMessage even on a
+// successful operation, which some servers use to signal warnings or degraded
+// state (for example, a directory in read-only / reinitializing mode).
+func ParseLDAPResult(packet *ber.Packet) (resultCode uint16, matchedDN, diagnosticMessage string) {
+	if packet == nil || len(packet.Children) < 2 || packet.Children[1] == nil {
+		return 0, "", ""
```
Comment (cpuschma) [inline v3/error.go]: A LDAPResult must have a `resultCode`, `matchedDN` and a `diagnosticMessage`. Of course the values can be empty except for resultCode, but only the `referral` is empty.

Silently discarding a malforrmed packet without returning an error can lead to unreliable behaviour.

File: v3/error.go [hunk 3]
```diff
@@ -195,6 +195,46 @@ func (e *Error) Error() string {
...
+	}
+	response := packet.Children[1]
+	if len(response.Children) < 1 || response.Children[0] == nil || response.Children[0].Value == nil {
+		return 0, "", ""
```
Comment (cpuschma) [inline v3/error.go]: Same goes here.

File: v3/error.go [hunk 4]
```diff
@@ -195,6 +195,46 @@ func (e *Error) Error() string {
...
+	}
+	switch v := response.Children[0].Value.(type) {
+	case int64:
+		resultCode = uint16(v)
+	case uint64:
+		resultCode = uint16(v)
+	case int:
+		resultCode = uint16(v)
+	default:
+		return 0, "", ""
```
Comment (cpuschma) [inline v3/error.go]: Same goes here.

**Review by cpuschma — CHANGES_REQUESTED** [review]
Please see the comments.

I would prefer not to make this function publicly available yet, until the aforementioned code snippets have been resolved. 

For now, please change the function to non-public, so that only the new fields in SearchResult are propagated.

File: v3/error.go (hunk 1 above)
Comment (Pujathacker2210) [reply v3/error.go]: Done — removed the uint64 and int cases, using a direct int64 assertion now

File: v3/error.go (hunk 2 above)
Comment (Pujathacker2210) [reply v3/error.go]: Good point. Changed to return an error describing the malformed packet.

File: v3/error.go (hunk 3 above)
Comment (Pujathacker2210) [reply v3/error.go]: Done — same approach, returns a descriptive error.

File: v3/error.go (hunk 4 above)
Comment (Pujathacker2210) [reply v3/error.go]: Done

**Comment by Pujathacker2210** [comment]
@cpuschma  Thanks for the review! Pushed the changes:

ParseLDAPResult → parseLDAPResult (unexported)
Simplified to int64 only per your note on TagInteger
All malformed-packet paths now return an error instead of silent zero values

Please take another look when you get a chance. Also, @t2y  raised two additional suggestions — adding DiagnosticMessage() to the Response interface and a DiagnosticMessage field to ldap.Error. Both would be breaking changes for external consumers. @cpuschma @johnweldon could you share your thoughts on those? Happy to incorporate them into this PR if there's consensus, or we can handle them as a follow-up.

File: v3/search.go [hunk 5]
```diff
@@ -595,16 +605,17 @@ func (l *Conn) Search(searchRequest *SearchRequest) (*SearchResult, error) {
...
-			}
 			entry := &Entry{
 				DN:         packet.Children[1].Children[0].Value.(string),
-				Attributes: attributes,
+				Attributes: unpackAttributes(attr),
 			}
 			result.Entries = append(result.Entries, entry)
 		case 5:
+			rc, _, diag, parseErr := parseLDAPResult(packet)
+			if parseErr == nil {
```
Comment (cpuschma) [inline v3/search.go]: `parseErr` is not returned, the error is silently dropped here.

**Review by cpuschma — CHANGES_REQUESTED** [review]

**Comment by cpuschma** [comment]
> @cpuschma Thanks for the review! Pushed the changes:
> 
> ParseLDAPResult → parseLDAPResult (unexported) Simplified to int64 only per your note on TagInteger All malformed-packet paths now return an error instead of silent zero values
> 
> Please take another look when you get a chance. Also, @t2y raised two additional suggestions — adding DiagnosticMessage() to the Response interface and a DiagnosticMessage field to ldap.Error. Both would be breaking changes for external consumers. @cpuschma @johnweldon could you share your thoughts on those? Happy to incorporate them into this PR if there's consensus, or we can handle them as a follow-up.

Let's not introduce breaking changes in a minor release. Currently we have a abundance of PRs regarding OOB reads for malformed packets we need to address first before new functions or even breaking changes can be introduced (#632)

File: v3/search.go [hunk 6]
```diff
@@ -601,6 +611,11 @@ func (l *Conn) Search(searchRequest *SearchRequest) (*SearchResult, error) {
 			}
 			result.Entries = append(result.Entries, entry)
 		case 5:
+			rc, _, diag, parseErr := parseLDAPResult(packet)
```
Comment (johnweldon) [inline v3/search.go]: Same point as @cpuschma's comment: `parseErr` is still ignored here as of 8ab56ea. A malformed SearchResultDone leaves `ResultCode` and `DiagnosticMessage` at their zero values, and the caller cannot tell.

File: v3/error.go [hunk 7]
```diff
@@ -195,6 +195,47 @@ func (e *Error) Error() string {
...
+	if response.Children[0] == nil || response.Children[0].Value == nil {
+		return 0, "", "", fmt.Errorf("malformed LDAPResult: nil resultCode")
+	}
+	v, ok := response.Children[0].Value.(int64)
+	if !ok {
+		return 0, "", "", fmt.Errorf("malformed LDAPResult: resultCode has unexpected type %T", response.Children[0].Value)
+	}
+	resultCode = uint16(v)
+	if response.Children[1] != nil {
+		if s, ok := response.Children[1].Value.(string); ok {
```
Comment (johnweldon) [inline v3/error.go]: A non-string matchedDN is accepted without an error, so `matchedDN` comes back as `""` with `err == nil`. Following @cpuschma's earlier comment, this should return a malformed-packet error the same way the resultCode checks do.

File: v3/error.go [hunk 8]
```diff
@@ -195,6 +195,47 @@ func (e *Error) Error() string {
...
+			matchedDN = s
+		}
+	}
+	if response.Children[2] != nil && response.Children[2].Value != nil {
+		if s, ok := response.Children[2].Value.(string); ok {
+			diagnosticMessage = s
+		} else {
+			diagnosticMessage = fmt.Sprintf("%v", response.Children[2].Value)
```
Comment (johnweldon) [inline v3/error.go]: A non-string diagnosticMessage is turned into text here. An INTEGER 7 comes back as `"7"` with no error. RFC 4511 defines it as an LDAPString, so this should be an error too. The same applies when `response.Children[2]` is nil.

File: v3/error.go [hunk 9]
```diff
@@ -195,6 +195,47 @@ func (e *Error) Error() string {
...
+// BER packet representing an LDAPResult (e.g. SearchResultDone).
+//
+// Unlike GetLDAPError, parseLDAPResult does not return early when resultCode is
+// success (0). This allows callers to read the diagnosticMessage even on a
+// successful operation, which some servers use to signal warnings or degraded
+// state (for example, a directory in read-only / reinitializing mode).
+//
+// Per RFC 4511, an LDAPResult must contain resultCode, matchedDN, and
+// diagnosticMessage. An error is returned if the packet is malformed.
+func parseLDAPResult(packet *ber.Packet) (resultCode uint16, matchedDN, diagnosticMessage string, err error) {
```
Comment (johnweldon) [inline v3/error.go]: Please add tests for this function: success with a diagnosticMessage, a non-zero resultCode, and each malformed case. `Search()` also needs tests showing that the new fields are set both on success and when it returns an error.

File: v3/search.go [hunk 10]
```diff
@@ -377,6 +377,14 @@ type SearchResult struct {
...
 	// Controls are the returned controls
 	Controls []Control
+	// ResultCode is the LDAP resultCode from SearchResultDone (0 = success).
+	// Populated even on success so callers can inspect it alongside DiagnosticMessage.
+	ResultCode uint16
+	// DiagnosticMessage is the server's free-text message from SearchResultDone.
+	// Per RFC 4511, this may be set even when resultCode is 0 (success).
+	// GetLDAPError returns nil on success and does not expose this field;
+	// it is preserved here so callers can log or act on it.
+	DiagnosticMessage string
```
Comment (johnweldon) [inline v3/search.go]: Because of `appendTo`, a paged search keeps the last page's `ResultCode` and `DiagnosticMessage`. `SearchAsync` does not fill these fields. Both facts belong in this doc comment.

**Review by johnweldon — CHANGES_REQUESTED** [review]
Thank you for your contribution!

**Comment by t2y** [comment]
I agree with @cpuschma's direction. Let's not include breaking changes in a minor release. We can revisit changing the interface later if needed.

File: v3/error.go (hunk 9 above)
Comment (Pujathacker2210) [reply v3/error.go]: Done — added TestParseLDAPResult with cases for success with/without diagnosticMessage, non-zero resultCode, and each malformed input.

File: v3/search.go (hunk 6 above)
Comment (Pujathacker2210) [reply v3/search.go]: Fixed — Search() now returns it wrapped in a NewError.

File: v3/search.go (hunk 10 above)
Comment (Pujathacker2210) [reply v3/search.go]: Added notes on both fields.

File: v3/error.go (hunk 7 above)
Comment (Pujathacker2210) [reply v3/error.go]: Fixed — returns a malformed-packet error now.

File: v3/error.go (hunk 8 above)
Comment (Pujathacker2210) [reply v3/error.go]: Fixed — returns a malformed-packet error now instead of converting with fmt.Sprintf.

File: v3/search.go (hunk 5 above)
Comment (Pujathacker2210) [reply v3/search.go]: Fixed
